### PR TITLE
Move the cell toolbar widget from the cell to the cell input

### LINKED ISSUE #13166: Side-by-Side rendering: cell toolbar disappears
## Description

Using Side-by-Side rendering, the cell toolbar might disappear in some situations:

https://user-images.githubusercontent.com/591645/193530736-40eccc9e-d36e-4608-b4d0-b666d0ce4a6c.mp4

## Reproduce

1. Create a new notebook
2. Add a new cell
3. Switch to `Side-by-side` rendering (`View > Render Side-by-Side`)
4. type `import this`, execute
5. The cell toolbar is gone for that cell

## Expected behavior

The cell toolbar should still be available when executing the stes described above.

## Context

- Operating System and version: Fedora 36
- Browser and version: Chrome 106
- JupyterLab version: 4.0.0a29

### LINKED ISSUE #15617: Cell toolbar position depends on cell header
## Description

I'm working on JupyterLab extension that will add widgets in cell header. I observe that when widgets are added to the cell header, then cell toolbar is moving up. It is expected to have cell toolbar always in the same position without dependency on cell header content.

#### Screenshot with header content
image

#### Screenshot without header content
image

## Reproduce

I attach code snippet that will create extension that will put two `div` elements to the cell header.

```ts
export class POCContentFactory extends NotebookPanel.ContentFactory {
  constructor(options: Cell.ContentFactory.IOptions) {
    super(options);
  }

  createCellHeader(): ICellHeader {
    return new POCCellHeader();
  }
}

export class POCCellHeader extends Widget implements ICellHeader {
  constructor() {
    super();
    this.layout = new PanelLayout();
    if (this.layout instanceof PanelLayout) {
      const t = document.createElement('div');
      t.textContent = 'Ola koduje';
      const a = new Widget({ node: t });
      this.layout.addWidget(a);
      const t2 = document.createElement('div');
      t2.textContent = 'Ola koduje';
      const b = new Widget({ node: t2 });
      this.layout.addWidget(b);
    }
  }
}

/* content factory extension for custom cell header */
const contentFactory: JupyterFrontEndPlugin<NotebookPanel.IContentFactory> = {
  id: 'piece-of-code-contend-factory-id',
  provides: NotebookPanel.IContentFactory,
  requires: [IEditorServices],
  autoStart: true,
  activate: (app: JupyterFrontEnd, editorServices: IEditorServices) => {
    const editorFactory = editorServices.factoryService.newInlineEditor;
    return new POCContentFactory({ editorFactory });
  }
};

```

## Expected behavior

Cell toolbar buttons position shouldn't depend on cell header.

I would be happy to provide PR with fix but I need assistance, where to look and where might be the issue. Maybe it is CSS issue?

### PR DESCRIPTION
**Pull request by brichet**
This PR changes the parent of the cell-toolbar, from the `Cell` widget to the `InputArea` widget.

## References

Fixes #15617
Fixes #13166 

As explained in https://github.com/jupyterlab/jupyterlab/issues/15402#issuecomment-1816055434 (end of comment), the cell toolbar is currently attached to the `Cell` widget. Its position is relative to the `Cell` dom, including the header.

Currently the cell header is not used, and has a height set to `0px`.

If the header get content (and height), the cell-toolbar position does not remain in the cell input.

## Code changes

Insert the `Toolbar` widget in the `Cell.inputArea` instead of the `Cell` itself, and modify the CSS accordingly.

## User-facing changes

Should not.

## Backwards-incompatible changes

As it changes the parent of the toolbar, it may break some extensions that rely on this assumption.
i.e. probably breaks https://github.com/jupyterlab-contrib/jlab-enhanced-cell-toolbar (yet not compatible with jupyterlab 4)

cc. @fcollonval

### CODE REVIEW

**Review by fcollonval — COMMENTED** [review]
The code looks good. But there is definitely side effect in tests.

**Comment by fcollonval** [comment]
Nicolas do you know if this will fix https://github.com/jupyterlab/jupyterlab/issues/13166?

**Comment by brichet** [comment]
> Nicolas do you know if this will fix #13166?

I didn't know this feature.
This fixes the problem, the toolbar is located in the input area. But if the input area is too short, the toolbar will disappear.

image

File: galata/test/jupyterlab/notebook-markdown.test.ts-snapshots/do-not-highlight-not-latex-jupyterlab-linux.png
Comment (krassowski) [inline galata/test/jupyterlab/notebook-markdown.test.ts-snapshots/do-not-highlight-not-latex-jupyterlab-linux.png:1]: This looks unexpected, I think the cell toolbar should not show up here

File: galata/test/jupyterlab/notebook-run.test.ts-snapshots/notebook-panel-1-jupyterlab-linux.png
Comment (krassowski) [inline galata/test/jupyterlab/notebook-run.test.ts-snapshots/notebook-panel-1-jupyterlab-linux.png:1]: Nor here.

File: galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png
Comment (krassowski) [inline galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png:1]: Nor here. It looks the logic for hiding toolbar no longer works with this change?

Comment (brichet) [reply galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png:1]: Thanks @krassowski for reviewing it.
Actually I don't know when the toolbar is supposed to hide, as soon as there is an active cell.
I did some tries with a released version of Jupyterlab and the toolbar never hides, even if the focus is not on the notebook panel.
Do you know how I can test toolbar hiding?

Comment (krassowski) [reply galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png:1]: > Actually I don't know when the toolbar is supposed to hide, as soon as there is an active cell.

The toolbar should hide if there is an overlap with the text in the cell, but indeed it does not always do that immediately after reloading (at least on `main`), which I believe is a regression, and needs fixing:

should-hide

Comment (krassowski) [reply galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png:1]: Of note this is a major source of test flakiness https://github.com/jupyterlab/jupyterlab/issues/15305

Comment (brichet) [reply galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png:1]: Thanks for the explication, this PR is actually breaking this logic.
I'll look at it and let you know.

**Comment by brichet** [comment]
I restored the logic to hide the toolbar, and reverted the modified snapshots.

File: galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png
Comment (brichet) [reply galata/test/documentation/overview.test.ts-snapshots/interface-jupyterlab-documentation-linux.png:1]: @krassowski the logic is fixed now

**Comment by krassowski** [comment]
Thanks for the ping @brichet. This looks good to me, can you just add an entry in https://github.com/jupyterlab/jupyterlab/blob/main/docs/source/extension/extension_migration.rst on this change and its implications?

**Comment by brichet** [comment]
This one should be ready now @krassowski.
Thanks again for the review.

**Review by krassowski — APPROVED** [review]
Thanks @brichet!

**Pull request merged**